packages/shadydom/src/patches/Element.js

```diff
@@ -120,6 +120,23 @@ export const ElementPatches = utils.getOwnPropertyDescriptors({
       this[utils.NATIVE_PREFIX + 'removeAttribute'](attr);
     }
   },
+
+  /**
+   * @this {Element}
+   * @param {string} attr
+   * @param {boolean | undefined} force
+   */
+  toggleAttribute(attr, force) {
+    if (this.ownerDocument !== doc) {
+      this[utils.NATIVE_PREFIX + 'toggleAttribute'](attr, force);
+    } else if (!scopeClassAttribute(this, attr, '')) {
+      this[utils.NATIVE_PREFIX + 'toggleAttribute'](attr, force);
+      distributeAttributeChange(this, attr);
+    } else if (this.getAttribute(attr) === '' && !force) {
+      // ensure that "class" attribute is fully removed if ShadyCSS does not keep scoping
+      this[utils.NATIVE_PREFIX + 'toggleAttribute'](attr, force);
+    }
```

Comment:
Is one of these cases guaranteed to happen?

Reply:
cc @sorvell

Reply:
Yeah, the code here is pretty messy unfortunately, and the test covereage is really poor. 

I would suggest just calling `this[utils.SHADY_PREFIX + 'set|removeAttribute'](...);` with the code you have in platform. (What is here is incorrect, at the least, because shadycss will always try to set the attribute.)

The shadydom tests are really spotty and messy but we should add something for this. Perhaps it could be dumped in here and modeled after a smaller subset of this test: https://github.com/webcomponents/polyfills/blob/4756d228c16a0f53a739846edeaebd740bb9dc09/packages/tests/shadydom/shady.html#L527.

Reply:
I think that implementing this in polyfill-style for toggleAttribute on top of setAttribute/removeAttribute also has a chance of being buggy, for which we need test coverage, and that if you distribute the cases in toggleAttribute through these set/removeAttribute methods you get to the logic here.

The main case is the middle one: `if (!scopeClassAttribute(this, attr, ''))` which happens if `attr` is _not_ class, and the same condition is used in both setAttribute and removeAttribute. The last case is only implemented in in removeAttribute, and so replicated here. The cases that fall though here, by inspection, also fall though in both setAttribute and removeAttribute.

If we think there are cases that might be missed and need to be covered by testing, I would like to add them to the tests regardless of implementation. I added versions of all tests I could find that already exist against set/remove and those pass.

Reply:
OK, that makes sense to me. Thanks for adding the extra tests.